src/huggingface_hub/snapshot_download.py

```diff
@@ -0,0 +1,62 @@
+import os
+from pathlib import Path
+from typing import Optional, Union
+
+from .constants import HUGGINGFACE_HUB_CACHE
+from .file_download import cached_download, hf_hub_url
+from .hf_api import HfApi
+
+
+REPO_ID_SEPARATOR = "__"
+# ^ make sure this substring is not allowed in repo_ids on hf.co
+
+
+def snapshot_download(
+    repo_id: str,
+    revision: Optional[str] = None,
+    cache_dir: Union[str, Path, None] = None,
+) -> str:
+    """
+    Downloads a whole snapshot of a repo's files at the specified revision.
+    This is useful when you want all files from a repo, because you don't know
+    which ones you will need a priori.
+    All files are nested inside a folder in order to keep their actual filename
+    relative to that folder.
+
+    An alternative would be to just clone a repo but this would require that
+    the user always has git and git-lfs installed, and properly configured.
+
+    Note: at some point maybe this format of storage should actually replace
+    the flat storage structure we've used so far (initially from allennlp
+    if I remember correctly).
+
+    Return:
+        Local folder path (string) of repo snapshot
+    """
+    if cache_dir is None:
+        cache_dir = HUGGINGFACE_HUB_CACHE
+    if isinstance(cache_dir, Path):
+        cache_dir = str(cache_dir)
+
+    _api = HfApi()
+    model_info = _api.model_info(repo_id=repo_id, revision=revision)
+
+    storage_folder = os.path.join(
+        cache_dir, repo_id.replace("/", REPO_ID_SEPARATOR) + "." + model_info.sha
+    )
+
+    for model_file in model_info.siblings:
+        url = hf_hub_url(
+            repo_id, filename=model_file.rfilename, revision=model_info.sha
+        )
+        relative_filepath = os.path.join(*model_file.rfilename.split("/"))
+
+        # Create potential nested dir
+        nested_dirname = os.path.dirname(
+            os.path.join(storage_folder, relative_filepath)
+        )
+        os.makedirs(nested_dirname, exist_ok=True)
+
+        cached_download(url, cache_dir=storage_folder, force_filename=relative_filepath)
```

Comment:
Something to note is that here we're not left with the exact state of the repo on the hub. We're left with three times as more files, as the `cached_download` also creates the metadata file (`xxx.yyy.json`) and the lock file (`xxx.yyy.lock`).

While this wasn't an issue previously, I'm afraid it becomes an issue here as we expect to have the repo exactly as it is on the hub and it's not the case.

I'm facing this issue in the AllenNLP integration, as when they detect there is a `vocabulary` directory, they try to load and parse *all* files within it.

I'm left with issues because where the code thought it was parsing `tokens.txt`, it is in fact parsing `tokens.txt.lock`. 

I think having these lock files is dangerous too, because downstream libraries might implement their own lockfile mechanisms and the most straightforward way to handle them is to have them be called `file.xxx.lock`, which would already exist in our case.

---

I see different ways of solving this issue, ordering them from least invasive to most invasive:

1. Remove metadata and lock files after download
2. Remove metadata and lock files *but* keep a global metadata file at the root, and document that
3. Leave it as it is, and have the libraries implement work-around themselves. 

I'd vote for 1. or 2.!

Reply:
Good point, and indeed I was wondering about this.

I think a variation of 1. is desired:
- in the `force_filename` mode, `cached_download` should not even store a metadata file (as its goal was mostly to be able to see which remote file a local file corresponds to, here we already know this)
- the `.lock` files we can (attempt to) remove in all cases (also regular `cached_download` operations)

What do you think?

Reply:
1. sounds good !

Moreover I the metadata were mainly used to get the orignal download url or the etag afaik, but here they are unique for each `repo_id`, `relative_filename` and  `revision` so we don't really need this in the end.